src/kirin/dialects/func/attrs.py

```diff
@@ -40,13 +40,18 @@ def print_impl(self, printer: Printer) -> None:
         printer.plain_print(" -> ")
         printer.print(self.output)
 
-    def serialize(self) -> dict[str, Any]:
-        # return the plain inner payload ONLY; global Serializer will encode nested
-        return {"inputs": [a for a in self.inputs], "output": self.output}
+    def serialize(self, serializer) -> dict[str, Any]:
```

Comment:
```suggestion
    def serialize(self, serializer) -> dict[str, Any]:
```

I think we should have either a base class or a protocal (pure type hint) for the `serializer`, we should not have any none-type-hinted code in this code base